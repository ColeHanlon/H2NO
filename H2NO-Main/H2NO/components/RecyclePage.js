import React from 'react'
import { View, Text, Button, Image, ScrollView, TouchableOpacity } from 'react-native'
import * as ImagePicker from 'expo-image-picker'
import { useState } from 'react';
import H2NOLogo from '../assets/H2NOLogo.png';

export default function RecyclePage({navigation}) {

    // Set default image
    const [image, setImage] = useState(require('../assets/H2NOLogo.png').uri);

    // Set image class
    const [objClass, setObjClass] = useState("~Classification will appear here!~");

    // Pick image from gallery
    const pickImage = async () => {
      let result = await ImagePicker.launchImageLibraryAsync({
        mediaTypes: ImagePicker.MediaTypeOptions.Images,
        aspect: [1,1],
        quality: 1,
        allowsEditing:true,
        base64: true,
      });
      if (!result.canceled) {
        setImage(result.assets[0].uri);
      
        var myHeaders = new Headers();
        myHeaders.append("Content-Type", "application/json");
        
        const imageb64 = `${result.assets[0].base64}`;
        var raw = JSON.stringify({"photo" : imageb64});
        var requestOptions = {
          method: 'POST',
          headers: myHeaders,
          body: raw,
          redirect: 'follow'
        };

        console.log('before post')
        setObjClass("Analyzing image!");

        fetch("http://192.168.86.244:5002/classify", requestOptions)
        .then(response => response.json())
        .then(result => setObjClass(result.Class))
        .catch(error => console.log('error', error));
        
        console.log('after post')
      }
    };

    // Take a photo
    const openCamera = async () => {
      // Ask the user for the permission to access the camera
      const permissionResult = await ImagePicker.requestCameraPermissionsAsync();
      
      if (permissionResult.granted === false) {
        alert("You have not allowed access to your camera!");
        return;
      }
      
      const result = await ImagePicker.launchCameraAsync(
        {aspect: [1,1],
        base64: true}
      );
      
      // Set the image that was taken
      if (!result.canceled) {
        setImage(result.assets[0].uri);
        var myHeaders = new Headers();
        myHeaders.append("Content-Type", "application/json");
        
        const imageb64 = `${result.assets[0].base64}`;
        var raw = JSON.stringify({"photo" : imageb64});
        var requestOptions = {
          method: 'POST',
          headers: myHeaders,
          body: raw,
          redirect: 'follow'
        };

        console.log('before post')
        setObjClass("Analyzing image!");
        
        fetch("http://192.168.86.244:5002/classify", requestOptions)
        .then(response => response.json())
        .then(result => setObjClass("That is a "+result.Class))
        .catch(error => console.log('error', error));

        console.log("after fetch");
      }
    }

    return (
      <ScrollView style={{backgroundColor: '#fdfdf6'}}>
      <View style={{flex: 1, alignItems:'center', justifyContent: 'center'}}>
        <Text></Text>
        <Text></Text>
        <Text style={{fontSize: 28, fontFamily: 'Cochin', textAlign: "center", marginBottom: 5}}>Upload an image to learn how to recycle it!</Text> 
        <Image source={image ? { uri: image } : H2NOLogo} style={{width:320, height: 320}} />
        <Button style={{fontFamily: 'Cochin'}} title="Select Photo to Upload" onPress={pickImage}/>
        <Button title="Take a Photo" onPress={openCamera}/>
        <Text></Text>
        <Text></Text>
        <Text style={{fontSize: 25, 
        fontFamily: 'Cochin', 
        textAlign: 'center'}}>{objClass}</Text>
      </View>
      <View style={{flex: 1, alignItems:'center', justifyContent: 'center', marginVertical: 50}}>
        <Text style={{fontsize: 100, fontFamily: 'Cochin'}}>Learn more about recycling! Click the button below</Text>
        <Text></Text>
        <TouchableOpacity
            onPress={() => navigation.navigate("Recycling 101")}
            style={{marginVertial: 10, backgroundColor: "gray", padding: 10, borderRadius: 15, width: 150, alignSelf: "center", backgroundColor: "#5ce697"}}>
            <Text style={{color: "white", textAlign: "center"}}>Recycling 101</Text>
        </TouchableOpacity>
      </View>
      </ScrollView>
    );
  }