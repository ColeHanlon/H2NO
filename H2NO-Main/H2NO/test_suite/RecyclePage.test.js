import React from 'react';
import renderer from 'react-test-renderer';
import RecyclePage from '../components/RecyclePage';

describe('<RecyclePage/>', () => {
  it("has 1 children", () => {
    const tree = renderer.create(<RecyclePage/>).toJSON();
    expect(tree.children.length).toBe(1);
  })
  it('renders correctly', () => {
    const tree = renderer.create(<RecyclePage/>).toJSON();
    expect(tree).toMatchSnapshot();
  })
});