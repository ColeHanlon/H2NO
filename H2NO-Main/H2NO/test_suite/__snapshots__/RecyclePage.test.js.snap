// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecyclePage/> renders correctly 1`] = `
<RCTScrollView
  style={
    Object {
      "backgroundColor": "#fdfdf6",
    }
  }
>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text />
      <Text />
      <Text
        style={
          Object {
            "fontFamily": "Cochin",
            "fontSize": 28,
            "marginBottom": 5,
            "textAlign": "center",
          }
        }
      >
        Upload an image to learn how to recycle it!
      </Text>
      <Image
        source={1}
        style={
          Object {
            "height": 320,
            "width": 320,
          }
        }
      />
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Select Photo to Upload
          </Text>
        </View>
      </View>
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Take a Photo
          </Text>
        </View>
      </View>
      <Text />
      <Text />
      <Text
        style={
          Object {
            "fontFamily": "Cochin",
            "fontSize": 25,
            "textAlign": "center",
          }
        }
      >
        ~Classification will appear here!~
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "marginVertical": 50,
        }
      }
    >
      <Text
        style={
          Object {
            "fontFamily": "Cochin",
            "fontsize": 100,
          }
        }
      >
        Learn more about recycling! Click the button below
      </Text>
      <Text />
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "#5ce697",
            "borderRadius": 15,
            "marginVertial": 10,
            "opacity": 1,
            "padding": 10,
            "width": 150,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "textAlign": "center",
            }
          }
        >
          Recycling 101
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
